src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,160 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String ERROR_MESSAGE_TEMPLATE = "Index: %d, Size: %d";
+
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddPossibility(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        unlink(node);
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if ((object == null && current.value == null)
+                    || (object != null && object.equals(current.value))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkAddPossibility(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(String.format(ERROR_MESSAGE_TEMPLATE, index, size));
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(String.format(ERROR_MESSAGE_TEMPLATE, index, size));
+        }
+    }
```

Comment:
Good idea, but basically those methods do the same job with same message template - so maybe be better to leave only one method?

Reply:
Yes, you're totally right, the methods are very similar, but have a different range in the loop itself 
`(index < 0 || index >= size)`
`(index < 0 || index > size)`. 
I don't even know how to get out of this situation...